### PR TITLE
React/ React Native udemy course

### PR DESCRIPTION
**Pull request by dwang134**
# Title: React/ React Native ZTM Section 03 
### Chapter 27: Class Components (changed to React Hooks with Typescript)
- changed react default function component syntax to class syntax
- import {Component} from 'react'
- class App extends Component{}
- render(){} with return ()

### Chapter 28: Component state (changed to useState hook)
- implemented constructor () method and called super() which calls the constructor the class extends from: Component
- instantiated class state with this.state {} object and created name property
- used { } braces to access value of this.state.name

### Chapter 29: onClick state hook
- added onClick property onto button and called setName function to update the name state
- React will only perform a repaint when a state is a different object in memory

### Chapter 30: states and shallow merge (Switched to React Hooks)
- shallow merge: look for the key in the current state obj and update them with the current one but will replace anything deeper than the first level
- to prevent shallow merge, when changing states you must ensure you change the object properties accordingly

### Chapter 31: states and secondary callback
- useState is an asynchronous hook, it will wait for the component to finish its cycle, re-render, and then it will update the state
- **useEffect() hook** replacing a call back for setState class syntax
- useEffect is a function that will be called everytime the component is rendered

### Chapter 32: Mapping Arrays to Elements
- App is the main component that will contain all our sub-components 
- implemented monsters state which is an array of Monsters {} object exported from types
- mapped through monsters array for each monster object

### Chapter 34: Keys for mapping
- Attached keys for each individual elements when mapping through it 
- Keys are used in React to identify which items in the list are changed, updated, or deleted

### Chapter 35:  non-SPA vs. SPA
non-SPA (pre-2010): when navigating to a non-SPA environment, your browser needs to request HTML, CSS, and JS from the server, when going into another division of the original URL your browser has to make that request from the server again
client --> makes a request for home page, server--> returns home page files

SPA: does that same thing as non-SPA, except this time the JS contains all the React-related code, and library, needed to build the website. And this time when you navigate to another part of the website under original URL, React will be responsible for serving HTML,CSS, and JS accordingly and not request from server again.

### Chapter 36: componentDidMount() (replaced by useEffect)
- componentDidMount triggers react code when the component has been mounted onto the DOM.
- main difference between useEffect and componentDidMount is that useEffect can happen asynchronously
- called an API and prototyped each object properties in TS

### Chapter 37: Promises
- can look at notes: https://docs.google.com/document/d/1PZo8exvfToyb7vAsjBOQ1spsUaEhkASVLqGrT-tP2PA/edit

### Chapter 38: Renders & Pre-renders
- chronological order of what is rendered first onto the screen when working with class components; constructor(), render()--> will run twice because state value changed, and then componentDidMount()
- same thing in React functional components context would be: useState, useEffect, and then the return ()

### Chapter 39: Input Search Box Component
- always implement functionality before CSS (recommended)
- we use classNames instead of class because our code is written in JSX; which means that class keyword is already taken (class component)
- input field onChange callback, e.target.value 

### Chapter 40: Searching & Filtering
- calling function that filters based on given targetValue from onChange
- has a problem because the code is directly changing the value of the state, and when re-typing values in the input box it filters from the result that has already been filtered

### Chapter 41: filter(), includes()
- filter(); creates a new array that doesn't match the given requirement
- includes(); filters result and check if it includes the given requirement 

### Chapter 42: Storing Original Data
- onChange property calls filter function to .toLowerCase(), filter data that .includes() targetValue, and then setting it to filteredMonsters state
- have to create a copy of the original data array, because when you're filtering you have to keep the original data to filter and not filter the data that is already filtered

### Chapter 43: Optimization
- in the course video, the implementation of the filter function is directly written inside of input onChange() property, which can slow down performance because it is an anonymous function and not stored anywhere. Which means everytime the state of the value is changed the onChange has to be re-intialized. 
- restructured class components at the top of render so "this" keyword can be avoided

### Chapter 44: Understanding Components
- Components ties together functionality and user representation into, aims to tie reusable portions of the code together into one body
- created components folder under src for implementation later on

### Chapter 45: CardList Component
- created separate folders for both CardList and Searchbox inside of src, so it's easier to bind related files (CSS) together (preference).
- each component return one element, or else it breaks DOM tree node

### Chapter 46: Component props
- going over how you can pass props to components in React
- props can be of any type desired

### Chapter 47: Rendering and Re-rendering
- render () components are rendered first, useEffect() is then called after the first render, which calls getUserData(), and calls both set functions for each states. Meaning if you put a console.log() inside of CardList component you will see it printing three times (once for first render, twice for both state values being changed).

### Chapter 48: Searchbox component 
- searchbox components configuration from top-level, passing filterData() function as a prop
- interfaced all expected props using TS

### Chapter 49: CSS in React
- talks about modularity between each CSS components, if you use filename.styles.css I think the file will be globalized across the entire project, also className declaration will be different

### Chapter 50: CardList Component
- generated unique images using unique ids from data array by inserting it into the HTTP link and using template literals
- styled card-container, individual card-list

### Chapter 51: Quick note on quotes for string interpolation
- javascript template literal syntax

### Chapter 52: Finishing touches
- adding more styling to make overall website look better

### Chapter 53: Functional components
- almost all enterprises uses functional components now vs class components 

### Chapter 54: class component life cycle
- https://projects.wojtekmaj.pl/react-lifecycle-methods-diagram/

### Chapter 55: Functional Components Intro
- return () replaces render() 
- need to understand what side-effects are because functional components life cycle differs from class component

### Chapter 56: Pure & Impure functions 
- https://www.educative.io/answers/pure-function-vs-impure-function
- pure function: function contains all the info that is needed already, don't need any variables or data coming from elsewhere
- impure function: function contains variables or data that needs to be properly managed elsewhere
- sideEffects(code that depends on an external factor): https://www.educative.io/answers/what-is-a-side-effect
- when using react hooks, you're generating side effects

### Chapter 57: Hooks: useState
- transforming class component this.state to functional component useState()

### Chapter 58: Functional Component re-rendering
- whenever value of state changes entire component gets re-runs, or re-render

### Chapter 59: Infinite re-rendering
- fetching from the API and using set state function to initialize the state value will cause the app the infinitely re-render
- this is caused because even though the array contains the exact same data the memory points to a different location so React recognizes that as a different array

### Chapter 60: useEffect()
- useEffect is a callback function that will be called after the components have been rendered
- second part of useEffect is where you insert dependencies, everytime the value of the dependencies have changed, the useEffect function will be called again
- you can create two useEffect() on one component and it will be ran in chronological order

### Chapter 61: Remaining Components
-converting the rest of the components from class component to functional component

### Chapter 62: Migrating from React v17 + ReactDOM v18 Changes
- upgrade react version in current package: npm upgrade react react-dom --latest
- new index.tsx
```
const root = ReactDOM.createRoot(
  document.getElementById('root') as HTMLElement
);
root.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>
);
```
### Chapter 63: Strict Mode Changes
- React StrictMode double renders everything so it can catch any weird behaviors or sideEffects inside your functional components (if you expected an output to print it twice, it will print out four times)
- download react dev tools, makes React Strictmode output greyed out so it's easier to differentiate

### Chapter 64: DOM and virtual DOM
- DOM tree: compose a structure of all the different elements in our application
- Monster rolodex DOM: App--> h1 title, Searchbox, CardList --> Cards
- Real DOM are expensive to make changes; re-flow
- React makes duplicate copy of real DOM but in JS: VDOM, significantly easier to do re-flow
- What happens if you type in searchbox: "lea"?  React creates a copy of VDOM, so we end up with VDOM copy and VDOM Snapshot. in When there is a value in the searchbox, it changes the value of the state variable residing in CardList, re-renders all the child and node elements of the App component, it also knows the only components it will need to update is the CardList, because the value of monster array state has changed, the props passed will also change, when new props are passed, React is also going to re-render that component, unmounts(remove) the child Card nodes matching the monster array.  React compares VDOM copy to snapshot and apply those changes directly onto the real DOM.

### Chapter 65-66: React and ReactDOM
- React library is the engine
- ReactDOM determines what we're trying to render to

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,23 +1,59 @@
-import React from 'react';
+import React, {useEffect, useState} from 'react';
 import logo from './logo.svg';
 import './App.css';
+import {User} from '../types/Monster'
+
+const App:React.FC= () =>  {
+  
+
+  //useState hooks
```
Comment (hsiangchiyang) [inline src/App.tsx]: Remove these lines. They are not needed.

File: src/App.tsx [hunk 2]
```diff
@@ -1,23 +1,59 @@
...
+
+
+    const getUserData = async() => {
+      const res = await fetch('https://jsonplaceholder.typicode.com/users');
+      const data = await res.json();
+      setMonsters(data);
+    }
+
+
+    // componentDidMount(
```
Comment (hsiangchiyang) [inline src/App.tsx]: Remove these lines.

File: types/Monster.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+
+//prototyping for Monster object
```
Comment (hsiangchiyang) [inline types/Monster.ts]: Remove the unused model.

**Review by hsiangchiyang — CHANGES_REQUESTED** [review]

**Review by dwang134 — COMMENTED** [review]
removing commented code resolved.

**Comment by hsiangchiyang** [comment]
Please merge this to main branch and start a new PR for the next topic.

**Pull request merged**